oak-run/src/main/java/org/apache/jackrabbit/oak/index/merge/IndexDiff.java

```diff
@@ -0,0 +1,434 @@
+                obj.getProperties().put(key, value);
+            }
+        }
+        for(String child : obj.getChildren().keySet()) {
+            JsonObject c = obj.getChildren().get(child);
+            sortPropertiesByName(c);
+        }
+    }
+
+    private static void compareIndexesIndexesInDirectory(Path indexPath, String index1, String index2,
+            JsonObject target) {
+        if (Files.isExecutable(indexPath)) {
+            indexFiles(indexPath).forEach(path -> {
+                JsonObject indexDefinitions = IndexDiff.parseIndexDefinitions(path.toString());
+                compareIndexes(indexDefinitions, indexPath.toString(), path.toString(), index1, index2, target);
+            });
+        } else {
+            JsonObject allIndexDefinitions = IndexDiff.parseIndexDefinitions(indexPath.toString());
+            for(String key : allIndexDefinitions.getChildren().keySet()) {
+                JsonObject indexDefinitions = allIndexDefinitions.getChildren().get(key);
+                compareIndexes(indexDefinitions, "", key, index1, index2, target);
+            }
+        }
+    }
+
+    private static void collectCustomizationsInDirectory(Path indexPath, JsonObject target) {
+        indexFiles(indexPath).forEach(path -> {
+            JsonObject indexDefinitions = IndexDiff.parseIndexDefinitions(path.toString());
+            showCustomIndexes(indexDefinitions, indexPath.toString(), path.toString(), target);
+        });
+    }
+
+    private static void mergeIndexesInDirectory(Path indexPath, JsonObject newIndex, JsonObject target) {
+        indexFiles(indexPath).forEach(path -> {
+            JsonObject indexDefinitions = IndexDiff.parseIndexDefinitions(path.toString());
+            mergeIndexes(indexDefinitions, indexPath.toString(), path.toString(), newIndex, target);
+        });
+    }
+
+    private static void mergeIndexes(JsonObject indexeDefinitions, String basePath, String fileName, JsonObject newIndexes, JsonObject target) {
+        JsonObject targetFile = new JsonObject(true);
+        if (newIndexes != null) {
+            for (String newIndexKey : newIndexes.getChildren().keySet()) {
+                if (indexeDefinitions.getChildren().containsKey(newIndexKey)) {
+                    targetFile.getProperties().put(newIndexKey, JsopBuilder.encode("WARNING: already exists"));
+                }
+            }
+        } else {
+            newIndexes = new JsonObject(true);
+        }
+        // the superseded indexes of the old repository
+        List<String> supersededKeys = new ArrayList<>(IndexMerge.getSupersededIndexDefs(indexeDefinitions));
+        Collections.sort(supersededKeys);
+
+        // keep only new indexes that are not superseded
+        Map<String, JsonObject> indexMap = indexeDefinitions.getChildren();
+        for (String superseded : supersededKeys) {
+            if (indexMap.containsKey(superseded)) {
+                indexMap.remove(superseded);
+            }
```

Comment:
the `containsKey` check could be removed

```suggestion
            indexMap.remove(superseded);
```

Reply:
+1